src/annotated/marked_yaml.rs

```diff
@@ -0,0 +1,152 @@
+//! A YAML node with position in the source document.
+//!
+//! This is set aside so as to not clutter `annotated.rs`.
+
+use hashlink::LinkedHashMap;
+use saphyr_parser::{Marker, Parser, ScanError};
+
+use crate::{LoadableYamlNode, Yaml, YamlData, YamlLoader};
+
+/// A YAML node with [`Marker`]s pointing to the start of the node.
+///
+/// This structure does not implement functions to operate on the YAML object. To access those,
+/// refer to the [`Self::data`] field.
+#[derive(Clone, Debug)]
+pub struct MarkedYaml {
+    /// The marker pointing to the start of the node.
+    ///
+    /// The marker is relative to the start of the input stream that was given to the parser, not
+    /// to the start of the document within the input stream.
+    pub marker: Marker,
+    /// The YAML contents of the node.
+    pub data: YamlData<MarkedYaml>,
+}
+
+impl MarkedYaml {
+    /// Load the given string as an array of YAML documents.
+    ///
+    /// See the function [`load_from_str`] for more details.
+    ///
+    /// # Errors
+    /// Returns `ScanError` when loading fails.
+    ///
+    /// [`load_from_str`]: `Yaml::load_from_str`
+    pub fn load_from_str(source: &str) -> Result<Vec<Self>, ScanError> {
+        Self::load_from_iter(source.chars())
+    }
+
+    /// Load the contents of the given iterator as an array of YAML documents.
+    ///
+    /// See the function [`load_from_str`] for more details.
+    ///
+    /// # Errors
+    /// Returns `ScanError` when loading fails.
+    ///
+    /// [`load_from_str`]: `Yaml::load_from_str`
+    pub fn load_from_iter<I: Iterator<Item = char>>(source: I) -> Result<Vec<Self>, ScanError> {
+        let mut parser = Parser::new(source);
+        Self::load_from_parser(&mut parser)
+    }
+
+    /// Load the contents from the specified [`Parser`] as an array of YAML documents.
+    ///
+    /// See the function [`load_from_str`] for more details.
+    ///
+    /// # Errors
+    /// Returns `ScanError` when loading fails.
+    ///
+    /// [`load_from_str`]: `Yaml::load_from_str`
+    pub fn load_from_parser<I: Iterator<Item = char>>(
+        parser: &mut Parser<I>,
+    ) -> Result<Vec<Self>, ScanError> {
+        let mut loader = YamlLoader::<Self>::default();
+        parser.load(&mut loader, true)?;
+        Ok(loader.into_documents())
+    }
+}
+
+impl PartialEq for MarkedYaml {
+    fn eq(&self, other: &Self) -> bool {
+        self.data.eq(&other.data)
+    }
+}
+
+// I don't know if it's okay to implement that, but we need it for the hashmap.
+impl Eq for MarkedYaml {}
```

Comment:
Most of the time, the offending types for `Eq` are types that can contain floats transitively (because this is the typical primitive value that isn't `Eq`). Which isn't the case here, so I believe it's entirely fine.